rmw_fastrtps_cpp/src/rmw_wait.cpp

```diff
@@ -148,21 +148,28 @@ rmw_wait(
   // If wait_timeout is not null and either of its fields are nonzero, we have to wait
   bool hasToWait = (wait_timeout && (wait_timeout->sec > 0 || wait_timeout->nsec > 0)) ||
     !wait_timeout;
-  hasToWait &= !check_wait_set_for_data(subscriptions, guard_conditions, services, clients);
+  bool hasData = check_wait_set_for_data(subscriptions, guard_conditions, services, clients);
+  hasToWait &= !hasData;
 
   bool timeout = false;
-
-  if (hasToWait) {
-    if (!wait_timeout) {
+  if (hasToWait && wait_timeout) {
+    auto predicate = [subscriptions, guard_conditions, services, clients]() {
+        return check_wait_set_for_data(subscriptions, guard_conditions, services, clients);
+      };
+    auto n = std::chrono::duration_cast<std::chrono::nanoseconds>(
+      std::chrono::seconds(wait_timeout->sec));
+    n += std::chrono::nanoseconds(wait_timeout->nsec);
+    timeout = !conditionVariable->wait_for(lock, n, predicate);
+  } else {
+    if (hasToWait && !wait_timeout) {
       conditionVariable->wait(lock);
```

Comment:
`std::condition_variable::wait(...)` is allowed to spuriously wake up. This could cause `rmw_wait()` to return with no data despite being told to wait forever. Would you mind including a fix for this by changing it to `conditionVariable->wait(lock, predicate)`?